pkg/cleaner/cleaner.go

```diff
@@ -0,0 +1,373 @@
+	if err != nil {
+		return err
+	}
+
+	log.Print("================ Remove pvc sub-path folder cleaner pod")
+
+	return cleaner.waitAndDeleteCleanUpPod(podName, "component="+podName, func(pvcSubpaths []*model.PVCSubPath) {}, []*model.PVCSubPath{})
+}
+
+func (cleaner *PVCSubPathCleaner) cleanUpSubPathFoldersContent() error {
+	pipelineRuns, err := cleaner.pipelineRunApi.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	pvcToCleanUp, err := cleaner.getPVCSubPathToContentCleanUp(pipelineRuns)
+	if err != nil {
+		return err
+	}
+
+	if len(pvcToCleanUp) == 0 {
+		log.Println("--------- Nothing to cleanup")
+		return nil
+	}
+
+	if isPVCSubPathCleanerRunning {
+		log.Println("--------- Skip pvc sub-path folder content cleaner, pvc sub-path folder cleaner is running.")
+		return nil
+	}
+
+	var delFoldersContentCmd string
+	var volumeMounts []corev1.VolumeMount
+	for _, pvc := range pvcToCleanUp {
+		delFoldersContentCmd += "cd /workspace/source/" + pvc.PVCSubPath + "; ls -A | xargs rm -rfv;"
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      "source",
+			MountPath: "/workspace/source/" + pvc.PVCSubPath,
+			SubPath:   pvc.PVCSubPath,
+		})
+	}
+
+	log.Println("--------- Create new pvc sub-path folder content cleaner pod")
+
+	pvcSubPathCleanerPod := cleaner.getPodCleaner("clean-pvc-sub-path-content-pod", "cleaner-pod", delFoldersContentCmd, volumeMounts, "registry.access.redhat.com/ubi8/ubi")
+	_, err = cleaner.clientset.CoreV1().Pods(cleaner.namespace).Create(context.TODO(), pvcSubPathCleanerPod, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	log.Println("--------- Remove pvc sub-path folder content cleaner pod")
+
+	return cleaner.waitAndDeleteCleanUpPod(pvcSubPathCleanerPod.Name, "component=cleaner-pod", cleaner.deletePVCFromStorage, pvcToCleanUp)
+}
+
+func (cleaner *PVCSubPathCleaner) waitAndDeleteCleanUpPod(podName string, label string, onDelete func([]*model.PVCSubPath), subPaths []*model.PVCSubPath) error {
+	watch, err := cleaner.clientset.CoreV1().Pods(cleaner.namespace).Watch(context.TODO(), metav1.ListOptions{
+		LabelSelector: label,
+	})
+	if err != nil {
+		return err
+	}
+
+	cleanUpDone := make(chan bool)
+	go func(cleanUpDone chan bool) {
+		for event := range watch.ResultChan() {
+			p, ok := event.Object.(*corev1.Pod)
+			if !ok {
+				continue
+			}
+			if p.Status.Phase == corev1.PodSucceeded {
+				log.Println("Pod cleaner succeeded")
+				cleanUpDone <- true
+			}
+			if p.Status.Phase == corev1.PodFailed {
+				log.Println("Pod cleaner failed" + p.Status.Reason + " " + p.Status.Message)
+				cleanUpDone <- false
+			}
+		}
+	}(cleanUpDone)
+
+	ticker := time.NewTicker(CLEANUP_TIMEOUT * time.Second)
+	for {
+		select {
+		case <-cleanUpDone:
+			ticker.Stop()
+			watch.Stop()
+			defer onDelete(subPaths)
+			return cleaner.clientset.CoreV1().Pods(cleaner.namespace).Delete(context.TODO(), podName, metav1.DeleteOptions{})
```

Comment:
Can we move this code after switch statement and left only warning in the second case?

Reply:
Sorry, I have no clue how to do that.

Reply:
I meant something like this:
```go
select {
	case <-cleanUpDone:
		fmt.Println("[INFO] Pod cleaner finished successfully")
	case <-ticker.C:
		fmt.Println("[WARN] Remove pod cleaner due timeout")
}
ticker.Stop()
watch.Stop()
defer onDelete(subPaths)
return cleaner.clientset.CoreV1().Pods(cleaner.namespace).Delete(context.TODO(), podName, metav1.DeleteOptions{})
```

Reply:
BTW, why do we need to wrap `select` in an infinite loop? If a signal comes, function returns anyway.